### ISSUE TITLE
Consider switching Interval parsing to a port of the Postgres code

### ISSUE THREAD

**Issue by jkosh44** [MaterializeInc/materialize#10613]
There have been various Interval parsing related issues fixed in the past, and some that are still not fixed (#10536, #10534, #10438, #7062). There's been some discussion around re-writing the parser as a direct port of the PostgreSQL Interval parser.

The Postgres parser can be found here: https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/timestamp.c#L875-L952
Some useful helper functions can be found here:
https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/datetime.c#L550-L775
https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/datetime.c#L3084-L3453
https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/datetime.c#L3500-L3713
https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/timestamp.c#L1993-L2007
https://github.com/postgres/postgres/blob/549ec201d6132b7c7ee11ee90a4e02119259ba5b/src/backend/utils/adt/timestamp.c#L1325-L1491

Porting over the Postgres parser would a bit of a large undertaking, so I thought I would open this issue up for discussion. Part of what makes it difficult is that the Postgres Interval parser uses a generic data date/time tokenizer and uses the Time parser. So we'd have to port both of those too.

**Comment by benesch** [comment]
> Porting over the Postgres parser would a bit of a large undertaking, so I thought I would open this issue up for discussion.

If you are up for it, do it! We need to solve this before we stabilize Platform, IMO. I'd even go so far as to say that we should parse into an interval type that directly matches the PostgreSQL interval type (with a separate "days" field), and only do the lossy conversion into our `repr::Interval` type at the very end. That will tee us up to support PostgreSQL intervals exactly when we have a path forward on #10566.

> Part of what makes it difficult is that the Postgres Interval parser uses a generic data date/time tokenizer and uses the Time parser. So we'd have to port both of those too.

That'd be great, because then we could make date/time parsing code match PostgreSQL too, yeah?

**Comment by jkosh44** [comment]
I agree. One question is, when we do the lossy conversion into our `repr::Interval` type, do you think there's a specific format that would be the best to prepare for #10566? Right now we store the total number of months as an i32 and the nanoseconds as an i128. 

Should we be collapsing the months into nanoseconds as well? That would make our `Interval` even more lossy, but it would allow us to compare `30 days` to `1 month`, which is what Postgres does.

**Comment by jkosh44** [comment]
I started porting over the generic date/time lexer on this branch: https://github.com/jkosh44/materialize/tree/pg-port. It's a bit rough right now but it seems to work somewhat well.
